File: BOJ/♠ 5000~5999/5530_JOIOI_탑.cpp
#include <iostream>
#include <vector>
using namespace std;

char joi_sample[1000001];
char joi[1000001];
int n;

bool is_possible(int num)
{
    for (int i = 0; i < n; i++)
        joi[i] = joi_sample[i];
    int left = num;

    for (int i = n - 1; i >= 0; i--)
    {
        if (left == 0)
            break;

        if (joi[i] == 'I')
        {
            joi[i] = 'E';
            left--;
        }
    }

    vector<int> a, b, c;
    for (int i = 0; i < n; i++)
    {
        if (joi[i] == 'E')
            c.push_back(i);
        else if (joi[i] == 'O')
            b.push_back(i);
        else
            a.push_back(i);
    }

    while (true)
    {
        if (c.size() == 0)
            return true;

        if (a.size() == 0 || b.size() == 0)
            return false;

        if (b.back() > c.back())
        {
            b.pop_back();
            continue;
        }

        if (a.back() > b.back())
        {
            a.pop_back();
            continue;
        }

        c.pop_back();
        b.pop_back();
        a.pop_back();
    }
}

int main()
{
    cin.tie()->sync_with_stdio(0);

    cin >> n;
    cin >> joi_sample;

    int start = 0;
    int end = 0;
    for (int i = 0; i < n; i++)
    {
        if (joi_sample[i] == 'I')
            end++;
    }

    int mid;
    while (start <= end)
    {
        mid = (start + end) / 2;

        if (is_possible(mid))
            start = mid + 1;
        else
            end = mid - 1;
    }

    cout << end;
}

File: BOJ/♠ 5000~5999/5530_JOIOI_탑.js
const joiSample = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n')[1];
let joi = new Array(joiSample.length);
let answer = 0;

function isPossible(num) {
    for (let i = 0; i < joiSample.length; i++)
        joi[i] = joiSample[i];
    left = num;

    for (let i = joi.length - 1; i >= 0; i--) {
        if (left === 0)
            break;

        if (joi[i] === 'I') {
            joi[i] = 'E';
            left--;
        }
    }

    let a = [];
    let b = [];
    let c = [];

    for (let i = 0; i < joi.length; i++) {
        if (joi[i] === 'E')
            c.push(i);
        else if (joi[i] === 'O')
            b.push(i);
        else
            a.push(i);
    }

    while (true) {
        if (c.length === 0)
            return true;

        if (a.length === 0 || b.length === 0)
            return false;

        if (b[b.length - 1] > c[c.length - 1]) {
            b.pop();
            continue;
        }

        if (a[a.length - 1] > b[b.length - 1]) {
            a.pop();
            continue;
        }

        c.pop();
        b.pop();
        a.pop();
    }
}

let start = 0;
let end = 0;
for (let i = 0; i < joiSample.length; i++) {
    if (joiSample[i] === 'I')
        end++;
}

let mid;
while (start <= end) {
    mid = Math.floor((start + end) / 2);

    if (isPossible(mid))
        start = mid + 1;
    else
        end = mid - 1;
}

console.log(end);